Source/Core/Common/FileSearch.cpp

```diff
@@ -25,119 +25,72 @@ namespace fs = std::filesystem;
 #endif
 
-namespace Common
-{
-#ifndef HAS_STD_FILESYSTEM
-
-static void FileSearchWithTest(const std::string& directory, bool recursive,
-                               std::vector<std::string>* result_out,
-                               std::function<bool(const File::FSTEntry&)> callback)
-{
-  File::FSTEntry top = File::ScanDirectoryTree(directory, recursive);
-
-  const std::function<void(File::FSTEntry&)> DoEntry = [&](File::FSTEntry& entry) {
-    if (callback(entry))
-      result_out->push_back(entry.physicalName);
-    for (auto& child : entry.children)
-      DoEntry(child);
-  };
-
-  for (auto& child : top.children)
-    DoEntry(child);
-}
+namespace fs = std::filesystem;
 
-std::vector<std::string> DoFileSearch(const std::vector<std::string>& directories,
-                                      const std::vector<std::string>& exts, bool recursive)
+namespace Common
 {
-  std::vector<std::string> result;
-
-  bool accept_all = exts.empty();
-  const auto callback = [&exts, accept_all](const File::FSTEntry& entry) {
-    if (accept_all)
-      return true;
-    if (entry.isDirectory)
-      return false;
-    return std::any_of(exts.begin(), exts.end(), [&](const std::string& ext) {
-      const std::string& name = entry.virtualName;
-      return name.length() >= ext.length() &&
-             strcasecmp(name.c_str() + name.length() - ext.length(), ext.c_str()) == 0;
-    });
-  };
-
-  for (const std::string& directory : directories)
-  {
-#ifdef ANDROID
-    // While File::ScanDirectoryTree (which is called in FileSearchWithTest) does handle Android
-    // content correctly, having a specialized implementation of DoFileSearch for Android content
-    // provides a much needed performance boost. Also, this specialized implementation will be
-    // required if we in the future replace the use of File::ScanDirectoryTree with std::filesystem.
-    if (IsPathAndroidContent(directory))
-    {
-      const std::vector<std::string> partial_result =
-          DoFileSearchAndroidContent(directory, exts, recursive);
-
-      result.insert(result.end(), std::make_move_iterator(partial_result.begin()),
-                    std::make_move_iterator(partial_result.end()));
-    }
-    else
-#endif
-    {
-      FileSearchWithTest(directory, recursive, &result, callback);
-    }
-  }
-
-  // remove duplicates
-  std::sort(result.begin(), result.end());
-  result.erase(std::unique(result.begin(), result.end()), result.end());
-  return result;
-}
-
-#else
-
 std::vector<std::string> DoFileSearch(const std::vector<std::string>& directories,
                                       const std::vector<std::string>& exts, bool recursive)
 {
-  bool accept_all = exts.empty();
+  const bool accept_all = exts.empty();
 
   std::vector<fs::path> native_exts;
   for (const auto& ext : exts)
     native_exts.push_back(StringToPath(ext));
 
   // N.B. This avoids doing any copies
   auto ext_matches = [&native_exts](const fs::path& path) {
```

Comment:
Untested but I think you could just use `extension` and avoid the platform specifics like this:

```
auto ext_matches = [&native_exts](const fs::path& path)
{
	return std::any_of(native_exts.cbegin(), native_exts.cend(), [&native_path](const auto& ext)
	{
		if (ext.length() != native_path.extension().length())
			return false;
		const std::string_view ext1 = native_path.extension().native();
		const std::string_view ext2 = ext.native();
		return std::equal(ext1.begin(), ext1.end(), ext2.begin(), [](auto c1, auto c2)
		{
			return std::tolower(c1) == std::tolower(c2);
		});
	});
};
```

Reply:
Hmm, might not handle wide characters :/

Reply:
What copies are being done here?

Reply:
Thanks for the detailed explanation.

I forgot that `extension` reports a copy.  I did see that there was a C++ proposal for a path view that hasn't been approved yet as far as I could tell.  Makes me wonder what `native()` is doing, I doubt it is returning a `string_view`.

Interesting to hear about the string comparison optimization, that makes sense.

Given all that and the fact that we're not saving much in terms of readability, I have no concerns.  Thanks again for the explanation and your patience with this @shuffle2 !